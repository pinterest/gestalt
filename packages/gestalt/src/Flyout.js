// @flow
import * as React from 'react';
import PropTypes from 'prop-types';
import Controller from './Controller.js';

type Props = {|
  anchor: ?HTMLElement,
  children?: React.Node,
  idealDirection?: 'up' | 'right' | 'down' | 'left',
  onDismiss: () => void,
  positionRelativeToAnchor?: boolean,
  color?: 'blue' | 'orange' | 'white',
  size?: 'xs' | 'sm' | 'md' | 'lg' | 'xl' | number,
|};

export default function Flyout(props: Props) {
  const {
    anchor,
    children,
    idealDirection,
    onDismiss,
    positionRelativeToAnchor = true,
    color = 'white',
    size,
  } = props;

  if (!anchor) {
    return null;
  }

  return (
    <Controller
      anchor={anchor}
      bgColor={color}
      idealDirection={idealDirection}
      onDismiss={onDismiss}
      positionRelativeToAnchor={positionRelativeToAnchor}
      shouldFocus={color !== 'orange'}
      size={size}
    >
      {children}
    </Controller>
  );
}

Flyout.propTypes = {
  anchor: PropTypes.shape({
    contains: PropTypes.func,
    getBoundingClientRect: PropTypes.func,
  }),
  children: PropTypes.node,
  idealDirection: PropTypes.oneOf(['up', 'right', 'down', 'left']),
  onDismiss: PropTypes.func.isRequired,
  positionRelativeToAnchor: PropTypes.bool,
  color: PropTypes.oneOf(['blue', 'orange', 'white']),
  size: PropTypes.oneOfType([
    PropTypes.number,
    PropTypes.oneOf(['xs', 'sm', 'md', 'lg', 'xl']), // default: sm
  ]),
};
