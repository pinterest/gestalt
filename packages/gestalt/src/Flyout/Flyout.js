// @flow
import * as React from 'react';
import PropTypes from 'prop-types';
import Controller from '../FlyoutUtils/Controller';

type Props = {|
  anchor: ?any,
  children?: any,
  idealDirection?: 'up' | 'right' | 'down' | 'left',
  onDismiss: () => void,
  positionRelativeToAnchor?: boolean,
  size?: 'xs' | 'sm' | 'md' | 'lg' | 'xl' | number,
|};

export default class Flyout extends React.PureComponent<Props> {
  render() {
    const {
      anchor,
      children,
      idealDirection,
      positionRelativeToAnchor = true,
      onDismiss,
      size,
    } = this.props;

    if (!anchor) {
      return null;
    }

    return (
      <Controller
        anchor={anchor}
        bgColor="white"
        idealDirection={idealDirection}
        onDismiss={onDismiss}
        positionRelativeToAnchor={positionRelativeToAnchor}
        shouldFocus
        size={size}
      >
        {children}
      </Controller>
    );
  }
}

Flyout.propTypes = {
  anchor: PropTypes.shape({
    contains: PropTypes.func,
    getBoundingClientRect: PropTypes.func,
  }),
  children: PropTypes.node,
  idealDirection: PropTypes.oneOf(['up', 'right', 'down', 'left']),
  onDismiss: PropTypes.func.isRequired,
  positionRelativeToAnchor: PropTypes.bool,
  size: PropTypes.oneOfType([
    PropTypes.number,
    PropTypes.oneOf(['xs', 'sm', 'md', 'lg', 'xl']), // default: sm
  ]),
};
