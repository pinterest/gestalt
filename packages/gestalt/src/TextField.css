.textField {
  composes: accessibilityOutline from "./Focus.css";
  composes: borderBox medium from "./Layout.css";
  composes: Text fontSize3 from "./Text.css";
  composes: xsCol12 from "./Column.css";
  appearance: none;
  border-radius: 4px;
  border-style: solid;
  border-width: 1px;
  padding: 8px 14px;
}

.textField::placeholder {
  color: #8e8e8e;
}

.textField:focus {
  background-color: #fff;
}

.normal {
  border-color: #8e8e8e;
}

input.errored {
  border-color: #e3780c;
}

input.errored:focus {
  box-shadow: 0 0 0 4px rgba(227, 120, 12, 0.75);
  outline: none;
}

.enabled {
  composes: darkGray from "./Colors.css";
  composes: whiteBg from "./Colors.css";
  cursor: text;
}

.disabled {
  composes: gray from "./Colors.css";
  composes: lightGrayBg from "./Colors.css";
}
