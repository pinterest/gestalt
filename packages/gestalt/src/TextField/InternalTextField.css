.textField {
  composes: borderBox from "../Layout.css";
  composes: Text from "../Text.css";
  composes: xsCol12 from "../Column.css";
}

.textField.sm {
  padding: var(--space-100) var(--space-200);
}

.textField.md {
  padding: var(--space-200) var(--space-300);
}

.textField.lg {
  padding: var(--space-300) var(--space-400);
}

.textField::placeholder {
  color: var(--color-text-formfield-placeholder);
}

.unstyledTextField,
.textFieldSpacer {
  composes: unstyled from "../sharedSubcomponents/FormElement.css";
  composes: Text from "../Text.css";
  composes: fontSize300 from "../Typography.css";
  margin-bottom: var(--space-100);
  padding: var(--space-100) var(--space-0);
}

.unstyledTextField {
  composes: absolute left0 top0 right0 bottom0 from "../Layout.css";
  width: 100%;
}

.unstyledTextField:focus {
  outline: 0;
}

.textFieldSpacer {
  visibility: hidden;
  white-space: nowrap;
}

.textFieldWrapper {
  display: flex;
  flex-wrap: wrap;
}

/*
Overriding the default bottom padding with important because tags need less padding at the bottom
*/

.sm.textFieldWrapper {
  /* The tag's margin bottom adds the other 2px */
  /* stylelint-disable-next-line declaration-no-important */
  padding-bottom: var(--space-0) !important;
  /* stylelint-disable-next-line declaration-no-important */
  padding-top: 2px !important;
}

.md.textFieldWrapper {
  /* The tag's margin bottom adds the other 2px */
  /* stylelint-disable-next-line declaration-no-important */
  padding-bottom: var(--space-100) !important;
  /* stylelint-disable-next-line declaration-no-important */
  padding-top: 6px !important;
}

.lg.textFieldWrapper {
  /* The larger tag's extra height and margin add the other 6px */
  /* stylelint-disable-next-line declaration-no-important */
  padding-bottom: var(--space-0) !important;
  /* stylelint-disable-next-line declaration-no-important */
  padding-top: 6px !important;
}

html:not([dir="rtl"]) .actionButton {
  padding-right: var(--space-800);
}

html[dir="rtl"] .actionButton {
  padding-left: var(--space-800);
}

.actionButtonContainer {
  bottom: var(--space-0);
  position: absolute;
  top: var(--space-0);
}

.vr_sm_actionButtonContainer {
  bottom: var(--sema-space-100);
  position: absolute;
}

.vr_md_actionButtonContainer {
  bottom: var(--sema-space-200);
  position: absolute;
}

.vr_lg_actionButtonContainer {
  bottom: var(--sema-space-300);
  position: absolute;
}

html:not([dir="rtl"]) .actionButtonContainer,
html:not([dir="rtl"]) .vr_sm_actionButtonContainer,
html:not([dir="rtl"]) .vr_md_actionButtonContainer,
html:not([dir="rtl"]) .vr_lg_actionButtonContainer {
  right: var(--space-0);
}

html[dir="rtl"] .actionButtonContainer,
html[dir="rtl"] .vr_sm_actionButtonContainer,
html[dir="rtl"] .vr_md_actionButtonContainer,
html[dir="rtl"] .vr_lg_actionButtonContainer {
  left: var(--space-0);
}
