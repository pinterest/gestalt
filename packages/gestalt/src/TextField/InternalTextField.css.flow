// @flow strict
/* This file is automatically generated by css-modules-flow-types */
declare module.exports: {|
  +'actionButton': string,
  +'actionButtonContainer': string,
  +'textField': string,
  +'textFieldSpacer': string,
  +'textFieldWrapper': string,
  +'unstyledTextField': string,
|};
