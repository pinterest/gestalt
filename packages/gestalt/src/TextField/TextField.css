.textField {
  composes: accessibilityOutline from "../Focus.css";
  composes: borderBox from "../Layout.css";
  composes: Text fontSize3 from "../Text/Text.css";
  composes: leadingShort from "../Typography.css";
  composes: xsCol12 from "../Column.css";
  appearance: none;
  border-radius: 4px;
  border-style: solid;
  border-width: 1px;
  max-height: 40px;
  padding: 10px 14px;
}

.textField::placeholder {
  color: #b5b5b5;
}

.textField:focus {
  background-color: #fff;
}

.normal {
  border-color: #b5b5b5;
}

.errored {
  border-color: #e3780c;
}

.enabled {
  composes: darkGray from "../Colors.css";
  composes: whiteBg from "../Colors.css";
  cursor: text;
}

.disabled {
  composes: gray from "../Colors.css";
  composes: lightGrayBg from "../Colors.css";
}
