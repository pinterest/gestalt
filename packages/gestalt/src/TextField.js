// @flow

import * as React from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';
import Flyout from './Flyout.js';
import Box from './Box.js';
import Text from './Text.js';
import styles from './TextField.css';

type State = {
  focused: boolean,
  errorIsOpen: boolean,
  errorMessage?: string,
};

type Props = {|
  autoComplete?: 'current-password' | 'on' | 'off' | 'username',
  disabled?: boolean,
  errorMessage?: string,
  hasError?: boolean,
  id: string,
  idealErrorDirection?: 'up' | 'right' | 'down' | 'left' /* default: right */,
  name?: string,
  onBlur?: ({
    event: SyntheticFocusEvent<HTMLInputElement>,
    value: string,
  }) => void,
  onChange: ({
    event: SyntheticInputEvent<HTMLInputElement>,
    value: string,
  }) => void,
  onFocus?: ({
    event: SyntheticFocusEvent<HTMLInputElement>,
    value: string,
  }) => void,
  onKeyDown?: ({
    event: SyntheticKeyboardEvent<HTMLInputElement>,
    value: string,
  }) => void,
  placeholder?: string,
  type?: 'date' | 'email' | 'number' | 'password' | 'text' | 'url',
  value?: string,
|};

export default class TextField extends React.Component<Props, State> {
  static propTypes = {
    autoComplete: PropTypes.oneOf([
      'current-password',
      'on',
      'off',
      'username',
    ]),
    disabled: PropTypes.bool,
    errorMessage: PropTypes.string,
    hasError: PropTypes.bool,
    id: PropTypes.string.isRequired,
    idealErrorDirection: PropTypes.string,
    name: PropTypes.string,
    onBlur: PropTypes.func,
    onChange: PropTypes.func.isRequired,
    onFocus: PropTypes.func,
    onKeyDown: PropTypes.func,
    placeholder: PropTypes.string,
    type: PropTypes.oneOf([
      'date',
      'email',
      'number',
      'password',
      'text',
      'url',
    ]),
    value: PropTypes.string,
  };

  static defaultProps = {
    disabled: false,
    hasError: false,
    idealErrorDirection: 'right',
    type: 'text',
  };

  state = {
    focused: false,
    errorIsOpen: false,
  };

  static getDerivedStateFromProps(props: Props, state: State) {
    if (props.errorMessage !== state.errorMessage) {
      return {
        errorIsOpen: !!props.errorMessage,
        errorMessage: props.errorMessage,
      };
    }

    return null;
  }

  handleChange = (event: SyntheticInputEvent<HTMLInputElement>) => {
    this.props.onChange({
      event,
      value: event.currentTarget.value,
    });
  };

  handleBlur = (event: SyntheticFocusEvent<HTMLInputElement>) => {
    if (this.props.errorMessage) {
      this.setState({ errorIsOpen: false });
    }
    if (this.props.onBlur) {
      this.props.onBlur({
        event,
        value: event.currentTarget.value,
      });
    }
  };

  handleFocus = (event: SyntheticFocusEvent<HTMLInputElement>) => {
    if (this.props.errorMessage) {
      this.setState({ errorIsOpen: true });
    }
    if (this.props.onFocus) {
      this.props.onFocus({
        event,
        value: event.currentTarget.value,
      });
    }
  };

  handleKeyDown = (event: SyntheticKeyboardEvent<HTMLInputElement>) => {
    if (this.props.onKeyDown) {
      this.props.onKeyDown({
        event,
        value: event.currentTarget.value,
      });
    }
  };

  textfield: ?HTMLElement;

  render() {
    const {
      autoComplete,
      disabled,
      errorMessage,
      hasError,
      id,
      idealErrorDirection,
      name,
      placeholder,
      type,
      value,
    } = this.props;

    const classes = classnames(
      styles.textField,
      disabled ? styles.disabled : styles.enabled,
      hasError || errorMessage ? styles.errored : styles.normal
    );

    // type='number' doesn't work on ios safari without a pattern
    // https://stackoverflow.com/questions/14447668/input-type-number-is-not-showing-a-number-keypad-on-ios
    const pattern = type === 'number' ? '\\d*' : undefined;

    return (
      <span>
        <input
          aria-describedby={
            errorMessage && this.state.focused ? `${id}-gestalt-error` : null
          }
          aria-invalid={errorMessage || hasError ? 'true' : 'false'}
          autoComplete={autoComplete}
          className={classes}
          disabled={disabled}
          id={id}
          name={name}
          onBlur={this.handleBlur}
          onChange={this.handleChange}
          onFocus={this.handleFocus}
          onKeyDown={this.handleKeyDown}
          pattern={pattern}
          placeholder={placeholder}
          ref={c => {
            this.textfield = c;
          }}
          type={type}
          value={value}
        />
        {errorMessage &&
          this.state.errorIsOpen && (
            <Flyout
              anchor={this.textfield}
              color="orange"
              idealDirection={idealErrorDirection}
              onDismiss={() => this.setState({ errorIsOpen: false })}
              size="sm"
            >
              <Box padding={3}>
                <Text bold color="white">
                  <span id={`${id}-gestalt-error`}>{errorMessage}</span>
                </Text>
              </Box>
            </Flyout>
          )}
      </span>
    );
  }
}
