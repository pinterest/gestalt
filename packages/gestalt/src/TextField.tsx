import { forwardRef, ReactElement, ReactNode, useEffect, useState } from 'react';
import { useDefaultLabelContext } from './contexts/DefaultLabelProvider';
import TagArea from './TagArea/TagArea';
import PasswordIconButton from './TextField/IconButtonEnd';
import InternalTextField, { autoCompleteType } from './TextField/InternalTextField';
import VRInternalTextField from './TextField/VRInternalTextField';
import useInExperiment from './useInExperiment';

export type MaxLength = {
  characterCount: number;
  errorAccessibilityLabel: string;
};

type Type = 'date' | 'email' | 'password' | 'tel' | 'text' | 'url';

type Props = {
  /**
     Indicate if autocomplete should be available on the input, and the type of autocomplete. All [autocomplete values](https://developer.mozilla.org/en-US/docs/Web/HTML/Attributes/autocomplete#values) are supported.
    */
  autoComplete?: autoCompleteType;
  /**
   * Indicate if the input is disabled. See the [disabled example](https://gestalt.pinterest.systems/web/textfield#Disabled) for more details.
   */
  disabled?: boolean;
  /**
   * Available for testing purposes, if needed. Consider [better queries](https://testing-library.com/docs/queries/about/#priority) before using this prop.
   */
  dataTestId?: string;
  /**
   * For most use cases, pass a string with a helpful error message (be sure to localize!). In certain instances it can be useful to make some text clickable; to support this, you may instead pass a React.Node to wrap text in [Link](https://gestalt.pinterest.systems/web/link) or [TapArea](https://gestalt.pinterest.systems/web/taparea).
   */
  errorMessage?: ReactNode;
  /**
   * This field is deprecated and will be removed soon. Please do not use.
   */
  hasError?: boolean;
  /**
   * More information about how to complete the form field.
   */
  helperText?: string;
  /**
   * A unique identifier for the input.
   */
  id: string;
  /**
   * The label for the input. Be sure to localize the text.
   */
  label?: string;
  /**
   * Whether the label should be visible or not. If `hidden`, the label is still available for screen reader users, but does not appear visually. See the [label visibility variant](https://gestalt.pinterest.systems#Label-visibility) for more info.
   */
  labelDisplay?: 'visible' | 'hidden';
  /**
   * The maximum number of characters allowed in Textfield. `maxLength` must be an integer value 0 or higher. See the [maximum length variant](https://gestalt.pinterest.systems/web/textfield#Maximum-length) for more details.
   */
  maxLength?: MaxLength;
  /**
   * Mobile only prop. Optionally specify the action label to present for the enter key on virtual keyboards. See the [enterKeyHint variant](https://gestalt.pinterest.systems/web/textfield#EnterKeyHint) for more info.
   */
  mobileEnterKeyHint?: 'enter' | 'done' | 'go' | 'next' | 'previous' | 'search' | 'send';
  /**
   * Mobile only prop. Optionally specify the type of data that might be entered by the user while editing the element or its contents. This allows mobile browsers to display an appropriate virtual keyboard. See the [enterKeyHint variant](https://gestalt.pinterest.systems/web/textfield#EnterKeyHint) for more info.
   */
  mobileInputMode?: 'none' | 'text' | 'decimal' | 'numeric';
  /**
   * A unique name for the input.
   */
  name?: string;
  /**
   * Callback triggered when the user blurs the input.
   */
  onBlur?: (arg1: { event: React.FocusEvent<HTMLInputElement>; value: string }) => void;
  /**
   * Callback triggered when the value of the input changes.
   */
  onChange: (arg1: { event: React.ChangeEvent<HTMLInputElement>; value: string }) => void;
  /**
   * Callback triggered when the user focuses the input.
   */
  onFocus?: (arg1: { event: React.FocusEvent<HTMLInputElement>; value: string }) => void;
  /**
   * Callback triggered when the user presses any key while the input is focused.
   */
  onKeyDown?: (arg1: { event: React.KeyboardEvent<HTMLInputElement>; value: string }) => void;
  /**
   * Placeholder text shown the the user has not yet input a value.
   */
  placeholder?: string;
  /**
   * Indicate if the input is readOnly. See the [readOnly example](https://gestalt.pinterest.systems/web/textfield#Read-only) for more details.
   */
  readOnly?: boolean;
  /**
   * Ref that is forwarded to the underlying input element.
   */
  ref?: ReactElement; // eslint-disable-line react/no-unused-prop-types,
  /**
   * List of tags to display in the component.
   */
  tags?: ReadonlyArray<ReactElement>;
  /**
   * The type of input. For non-telephone numerical input, please use [NumberField](https://gestalt.pinterest.systems/web/numberfield).
   */
  type?: Type;
  /**
   * Defines the height of the TextField: sm: 32px, md: 40px (default), lg: 48px. See the [size variant](https://gestalt.pinterest.systems/web/TextField#Size) for more details.
   */
  size?: 'sm' | 'md' | 'lg';
  /**
   * The current value of the input.
   */
  value?: string;
};

/**
 * [TextField](https://gestalt.pinterest.systems/web/textfield) allows for multiple types of text input.
 *
 * ![TextField light mode](https://raw.githubusercontent.com/pinterest/gestalt/master/playwright/visual-test/TextField.spec.ts-snapshots/TextField-chromium-darwin.png)
 * ![TextField dark mode](https://raw.githubusercontent.com/pinterest/gestalt/master/playwright/visual-test/TextField-dark.spec.ts-snapshots/TextField-dark-chromium-darwin.png)
 *
 */

const TextFieldWithForwardRef = forwardRef<HTMLInputElement, Props>(function TextField(
  {
    autoComplete,
    dataTestId,
    disabled = false,
    errorMessage,
    hasError = false,
    helperText,
    id,
    label,
    labelDisplay = 'visible',
    maxLength,
    mobileEnterKeyHint,
    mobileInputMode,
    name,
    onBlur,
    onChange,
    onFocus,
    onKeyDown,
    placeholder,
    readOnly = false,
    size = 'md',
    tags,
    type: typeProp = 'text',
    value,
  }: Props,
  ref,
) {
  /**
   * Yes, this is initializing a state variable with a prop value and then disregarding the prop value â€” often a code smell, I know. This is necessary to internalize the effective input type (password vs text) and not force the user to handle responding to clicks on the button
   */
  const [type, setType] = useState<Type>(typeProp);

  const isInVRExperiment = useInExperiment({
    webExperimentName: 'web_gestalt_visualRefresh',
    mwebExperimentName: 'web_gestalt_visualRefresh',
  });

  useEffect(() => {
    setType(typeProp);
  }, [typeProp]);

  const isPasswordField = typeProp === 'password';
  const isCurrentlyPasswordType = type === 'password';

  const { accessibilityHidePasswordLabel, accessibilityShowPasswordLabel } =
    useDefaultLabelContext('TextField');

  const iconButton = isPasswordField ? (
    <PasswordIconButton
      accessibilityChecked={!isCurrentlyPasswordType}
      accessibilityLabel={
        isCurrentlyPasswordType
          ? accessibilityShowPasswordLabel ?? ''
          : accessibilityHidePasswordLabel ?? ''
      }
      icon={isCurrentlyPasswordType ? 'eye' : 'eye-hide'}
      onClick={() => {
        setType(isCurrentlyPasswordType ? 'text' : 'password');
      }}
      paddingSize={size}
      role="switch"
      tooltipText={
        isCurrentlyPasswordType
          ? accessibilityShowPasswordLabel ?? ''
          : accessibilityHidePasswordLabel ?? ''
      }
    />
  ) : undefined;

  if (isInVRExperiment && !tags) {
    return (
      <VRInternalTextField
        ref={ref}
        autoComplete={autoComplete}
        dataTestId={dataTestId}
        disabled={disabled}
        errorMessage={errorMessage}
        hasError={hasError}
        helperText={helperText}
        iconButton={iconButton}
        id={id}
        label={label}
        labelDisplay={labelDisplay}
        maxLength={maxLength}
        mobileEnterKeyHint={mobileEnterKeyHint}
        mobileInputMode={mobileInputMode}
        name={name}
        onBlur={onBlur}
        onChange={onChange}
        onFocus={onFocus}
        onKeyDown={onKeyDown}
        placeholder={placeholder}
        readOnly={readOnly}
        size={size}
        tags={tags}
        type={type}
        value={value}
      />
    );
  }

  if (isInVRExperiment && tags)
    return (
      <TagArea
        // @ts-expect-error - TS2322
        ref={ref}
        dataTestId={dataTestId}
        disabled={disabled}
        errorMessage={errorMessage}
        hasError={hasError}
        helperText={helperText}
        id={id}
        label={label}
        labelDisplay={labelDisplay}
        maxLength={maxLength}
        name={name}
        onBlur={onBlur}
        onChange={onChange}
        onFocus={onFocus}
        onKeyDown={onKeyDown}
        placeholder={placeholder}
        readOnly={readOnly}
        size={size}
        tags={tags}
        value={value}
      />
    );

  return (
    <InternalTextField
      ref={ref}
      autoComplete={autoComplete}
      dataTestId={dataTestId}
      disabled={disabled}
      errorMessage={errorMessage}
      hasError={hasError}
      helperText={helperText}
      iconButton={iconButton}
      id={id}
      label={label}
      labelDisplay={labelDisplay}
      maxLength={maxLength}
      mobileEnterKeyHint={mobileEnterKeyHint}
      mobileInputMode={mobileInputMode}
      name={name}
      onBlur={onBlur}
      onChange={onChange}
      onFocus={onFocus}
      onKeyDown={onKeyDown}
      placeholder={placeholder}
      readOnly={readOnly}
      size={size}
      tags={tags}
      type={type}
      value={value}
    />
  );
});

TextFieldWithForwardRef.displayName = 'TextField';

export default TextFieldWithForwardRef;
