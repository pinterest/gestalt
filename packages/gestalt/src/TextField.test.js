// @flow
import React from 'react';
import { create } from 'react-test-renderer';
import { shallow } from 'enzyme';
import TextField from './TextField.js';
import Flyout from './Flyout.js';

describe('TextField', () => {
  it('Renders a Flyout if an error message is passed in', () => {
    const wrapper = shallow(
      <TextField errorMessage="test" id="test" onChange={jest.fn()} />
    );
    wrapper.instance().setState({ errorIsOpen: true });
    wrapper.simulate('focus');
    expect(wrapper.find(Flyout)).toHaveLength(1);
  });

  it('Does not render a Flyout when errorMessage is null', () => {
    const wrapper = shallow(<TextField id="test" onChange={jest.fn()} />);
    expect(wrapper.find(Flyout)).toHaveLength(0);
  });

  it('TextField normal', () => {
    const tree = create(
      <TextField
        id="test"
        onChange={jest.fn()}
        onFocus={jest.fn()}
        onBlur={jest.fn()}
      />
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('TextField with error', () => {
    const tree = shallow(
      <TextField
        errorMessage="error message"
        id="test"
        onChange={jest.fn()}
        onFocus={jest.fn()}
        onBlur={jest.fn()}
      />
    ).html();
    expect(tree).toMatchSnapshot();
  });

  it('TextField with hasError', () => {
    const wrapper = shallow(
      <TextField
        hasError
        id="test"
        onChange={jest.fn()}
        onFocus={jest.fn()}
        onBlur={jest.fn()}
      />
    );

    expect(wrapper.find(Flyout)).toHaveLength(0);
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('TextField with name', () => {
    const tree = shallow(
      <TextField
        name="email"
        id="email"
        onChange={jest.fn()}
        onFocus={jest.fn()}
        onBlur={jest.fn()}
      />
    ).html();
    expect(tree).toMatchSnapshot();
  });

  it('TextField with autocomplete', () => {
    const tree = shallow(
      <TextField
        autoComplete="on"
        id="email"
        onChange={jest.fn()}
        onFocus={jest.fn()}
        onBlur={jest.fn()}
      />
    ).html();
    expect(tree).toMatchSnapshot();
  });

  it('TextField with errorMessage prop change', () => {
    const tree = shallow(
      <TextField
        id="test"
        onChange={jest.fn()}
        onFocus={jest.fn()}
        onBlur={jest.fn()}
      />
    );
    expect(tree.find(Flyout)).toHaveLength(0);
    tree.setProps({
      errorMessage: 'error message',
    });
    expect(tree.find(Flyout)).toHaveLength(1);
  });

  it('TextField with type number', () => {
    const tree = shallow(
      <TextField
        id="test"
        onChange={jest.fn()}
        onFocus={jest.fn()}
        onBlur={jest.fn()}
        type="number"
      />
    ).html();
    expect(tree).toMatchSnapshot();
  });

  it('handles blur events', () => {
    const mockBlur = jest.fn();
    const tree = shallow(
      <TextField id="test" onBlur={mockBlur} onChange={jest.fn()} />
    );
    tree
      .find('input')
      .simulate('blur', { currentTarget: { value: 'fake value' } });
    expect(mockBlur).toHaveBeenCalledWith({
      event: { currentTarget: { value: 'fake value' } },
      value: 'fake value',
    });
  });

  it('handles change events', () => {
    const mockChange = jest.fn();
    const tree = shallow(<TextField id="test" onChange={mockChange} />);
    tree
      .find('input')
      .simulate('change', { currentTarget: { value: 'fake value' } });
    expect(mockChange).toHaveBeenCalledWith({
      event: { currentTarget: { value: 'fake value' } },
      value: 'fake value',
    });
  });

  it('handles focus events', () => {
    const mockFocus = jest.fn();
    const tree = shallow(
      <TextField id="test" onChange={jest.fn()} onFocus={mockFocus} />
    );
    tree
      .find('input')
      .simulate('focus', { currentTarget: { value: 'fake value' } });
    expect(mockFocus).toHaveBeenCalledWith({
      event: { currentTarget: { value: 'fake value' } },
      value: 'fake value',
    });
  });

  it('handles key down events', () => {
    const mockKeyDown = jest.fn();
    const tree = shallow(
      <TextField id="test" onChange={jest.fn()} onKeyDown={mockKeyDown} />
    );
    tree
      .find('input')
      .simulate('keyDown', { currentTarget: { value: 'fake value' } });
    expect(mockKeyDown).toHaveBeenCalledWith({
      event: { currentTarget: { value: 'fake value' } },
      value: 'fake value',
    });
  });
});
