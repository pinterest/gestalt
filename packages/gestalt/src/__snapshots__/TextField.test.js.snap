// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextField TextField normal 1`] = `
<span>
  <input
    aria-describedby={null}
    aria-invalid="false"
    className="textField enabled normal"
    disabled={false}
    id="test"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    type="text"
  />
</span>
`;

exports[`TextField TextField with autocomplete 1`] = `"<span><input type=\\"text\\" aria-invalid=\\"false\\" autoComplete=\\"on\\" class=\\"textField enabled normal\\" id=\\"email\\"/></span>"`;

exports[`TextField TextField with error 1`] = `"<span><input type=\\"text\\" aria-invalid=\\"true\\" class=\\"textField enabled errored\\" id=\\"test\\"/></span>"`;

exports[`TextField TextField with hasError 1`] = `"<span><input type=\\"text\\" aria-invalid=\\"true\\" class=\\"textField enabled errored\\" id=\\"test\\"/></span>"`;

exports[`TextField TextField with name 1`] = `"<span><input type=\\"text\\" aria-invalid=\\"false\\" class=\\"textField enabled normal\\" id=\\"email\\" name=\\"email\\"/></span>"`;

exports[`TextField TextField with type number 1`] = `"<span><input type=\\"number\\" aria-invalid=\\"false\\" class=\\"textField enabled normal\\" id=\\"test\\" pattern=\\"\\\\d*\\"/></span>"`;
